battery/imx6ull-vbat/imx6ull-vbat.c

```diff
@@ -0,0 +1,446 @@
+	uint32_t HS;
+	uint32_t R0;
+	uint32_t reserved1;
+	uint32_t CFG;
+	uint32_t GC;
+	uint32_t GS;
+	uint32_t CV;
+	uint32_t OFS;
+	uint32_t CAL;
+} adc_t;
+
+static struct {
+	gpio_t *gpio1;
+	gpio_t *gpio5;
+	adc_t *adc;
+	uint32_t port;
+	int busy;
+	char buf[16];
+	int len;
+} common;
+
+
+static int gpio_init(void)
+{
+	platformctl_t pctl;
+
+	common.gpio1 = mmap(NULL, 0x1000, PROT_READ | PROT_WRITE, MAP_DEVICE | MAP_PHYSMEM | MAP_ANONYMOUS, -1, GPIO1_BASE_ADDR);
+	if (common.gpio1 == MAP_FAILED) {
+		log("GPIO1 mmap failed");
+		return -1;
+	}
+
+	common.gpio5 = mmap(NULL, 0x1000, PROT_READ | PROT_WRITE, MAP_DEVICE | MAP_PHYSMEM | MAP_ANONYMOUS, -1, GPIO5_BASE_ADDR);
+	if (common.gpio5 == MAP_FAILED) {
+		log("GPIO5 mmap failed");
+		munmap((void *)common.gpio1, 0x1000);
+		return -1;
+	}
+
+	/* Configure VBAT pin as GPIO */
+	pctl.action = pctl_set;
+	pctl.type = pctl_iomux;
+	pctl.iomux.sion = 0;
+	pctl.iomux.mux = VBAT_GPIO_MUX;
+	pctl.iomux.mode = 5;
+	platformctl(&pctl);
+
+	/* Configure VBAT pad */
+	pctl.action = pctl_set;
+	pctl.type = pctl_iopad;
+	pctl.iopad.pad = VBAT_GPIO_PAD;
+	pctl.iopad.hys = 0;
+	pctl.iopad.pus = 0;
+	pctl.iopad.pue = 0;
+	pctl.iopad.pke = 0;
+	pctl.iopad.ode = 0;
+	pctl.iopad.speed = 2;
+	pctl.iopad.dse = 0;
+	pctl.iopad.sre = 0;
+	platformctl(&pctl);
+
+	/* Configure VBAT as input */
+	common.gpio1->GDIR &= ~(1U << VBAT_GPIO_PIN);
+
+	/* Configure VBAT_ON as output */
+	common.gpio5->GDIR |= 1U << VBAT_ON_GPIO_PIN;
+
+	/* Set VBAT_ON to 0 */
+	common.gpio5->DR &= ~(1U << VBAT_ON_GPIO_PIN);
+
+	return 0;
+}
+
+
+static void gpio_deinit(void)
+{
+	munmap((void *)common.gpio1, 0x1000);
+	munmap((void *)common.gpio5, 0x1000);
+}
+
+
+static int adc_init(void)
+{
+	platformctl_t pctl;
+
+	common.adc = mmap(NULL, 0x1000, PROT_READ | PROT_WRITE, MAP_DEVICE | MAP_PHYSMEM | MAP_ANONYMOUS, -1, ADC_BASE_ADDR);
+	if (common.adc == MAP_FAILED) {
+		log("ADC mmap failed");
+		return -1;
+	}
+
+	/* Enable ADC clock */
+	pctl.action = pctl_set;
+	pctl.type = pctl_devclock;
+	pctl.devclock.dev = ADC_CLK;
+	pctl.devclock.state = 3;
+	platformctl(&pctl);
+
+	/* Sample time duration is 20 ADC clocks, long sample time, 12-bit conversion mode */
+	common.adc->CFG = (2U << 8) | (1U << 4) | (2U << 2);
+
+	/* Start calibration */
+	common.adc->GC |= 1U << 7;
+
+	/* Wait until calibration is completed */
+	while ((common.adc->GC & (1U << 7)) != 0)
+		;
+
+	/* Check if calibration was successful */
+	if ((common.adc->GS & (1 << 1)) != 0) {
+		log("ADC calibration failed");
+		munmap((void *)common.adc, 0x1000);
+		return -1;
+	}
+
+	return 0;
+}
+
+
+static void adc_deinit(void)
+{
+	platformctl_t pctl;
+
+	munmap((void *)common.adc, 0x1000);
+
+	/* Disable ADC clock */
+	pctl.action = pctl_set;
+	pctl.type = pctl_devclock;
+	pctl.devclock.dev = ADC_CLK;
+	pctl.devclock.state = 0;
+	platformctl(&pctl);
+}
+
+
+static uint16_t adc_read(void)
+{
+	uint16_t value;
+
+	/* Set VBAT_ON to 1 */
+	common.gpio5->DR |= 1U << VBAT_ON_GPIO_PIN;
```

Comment:
Fuction is adc_read, but we use GPIO here too. This functionality should be in a different function or function should be renamed